src/racer/cargo.rs

```diff
@@ -518,8 +518,32 @@ fn getstr(t: &toml::Table, k: &str) -> Option<String> {
 fn find_cargo_tomlfile<P>(file: P) -> Option<PathBuf>
     where P: Into<PathBuf>
 {
-    find_next_crate_root(file)
-        .map(|mut f| { f.push("Cargo.toml"); f })
+    find_next_crate_root(file).map(|mut f| {
+        f.push("Cargo.toml");
+        f
+    })
+}
+
+/// Find workspace root by traversing up the directory tree
+///
+/// Checks every Cargo.toml found while searching up the directory tree to see if it has a
+/// [workspace] section in its top level.
+fn find_workspace_root<P>(path: P) -> Option<PathBuf>
+    where P: Into<PathBuf>
+{
+    let mut path = path.into();
+    path.push("Cargo.toml");
+    trace!("find_workspace_root checking {:?}", path);
+    if path.exists() &&
+       parse_toml_file(&path).as_ref().and_then(|toml| toml.get("workspace")).is_some() {
+        trace!("find_workspace_root found workspace section in {:?}", path);
+        path.pop();
+        Some(path)
+    } else if path.pop() && path.pop() {
+        find_workspace_root(path)
```

Comment:
Can you add a comment here that mentions this is discarding `Cargo.toml`, moving up the directory tree, and trying again?